README.md
![work-break](screenshot.png)

## Introduction
This balancer can track your work time and suggest break time. It uses this [graph](https://www.desmos.com/calculator/duqezlkow8) by default,
where the horizontal axis is for work time, and the vertical axis is for break time. It passes through the [Pomodoro Technique](https://en.wikipedia.org/wiki/Pomodoro_Technique) (25/5) and the [52/17 rule](https://en.wikipedia.org/wiki/52/17_rule). The formula can be adjusted (see the configuration section below).

The more you work, the more break you need per work minute. Sometimes, you have to start working without waiting until you have rested suggested time. In this case, this balancer accumulates strain by subtracting the actual break time from the needed break time and converting the result into work time back. It adds up to the following work on the next break.

This balancer sends you notifications on the following events:
* Work lasted for 25 minutes
* It lasted for 52 minutes
* It lasted for the daily work time limit if it configured
* Break ended

The notification contains the current phase, strain, needed break, its end if it starts, and work time today.

## How to install

### Compatibility

Windows 8+ (derived from dependencies, tested on Windows 10)

MacOS 10+ (derived from dependencies, not tested)

Linux/BSD: (should work with many distributions, tested on Arch Linux)

### Download an executable

Download a suitable executable from [releases](https://github.com/ShadoySV/work-break/releases) and copy it to some directory that lists in your PATH.

### With Cargo

You will need the Rust programming language and its cargo package manager installed on your system. See the official documentation [here](https://www.rust-lang.org/tools/install).

To install it on Windows, see the requirement [here](https://github.com/BenjaminRi/winresource#compiling-on-windows).

Run this command to install the app:
```
cargo install work-break
```

## How to use

Configure the following command to launch the balancer on startup to get notifications:
```
work-break
```

Configure a shortcut key or create a desktop icon to switch between work and break time:
```
work-break switch
```

Configure a shortcut key or create a desktop icon to notify about the current status:
```
work-break notify
```

## Configuration

After the first app launch, you can change the formula's **(break = a * (work ^ b) + c)** coefficients (work and break compute in seconds), daily work time limit (in minutes, zero turns the notification off), and hour when work days start at (from 0 to 23) by editing the following configuration file, or you can delete the file to get defaults:

##### Windows
```
%APPDATA%\work-break\config\default-config.toml
```

##### Unix
```
~/.config/work-break/default-config.toml
```

Then restart your system or ask the app to apply the configuration:

```
work-break reload
```

# Similar projects

[Bartib](https://github.com/nikolassv/bartib) is a simple time tracker for the command line. It saves a log of all tracked activities as a plaintext file and allows you to create flexible reports.


CHANGELOG
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.1] - 2023-08-12

### Added

- Using work-break icon (if exists in system, Linux/BSD)

## [0.2.0] - 2023-07-23

### Added

- Cross-platform support using Desktop notifications
- Counting the total work time of the day
- Notifying on configured daily work time limit
- Ability to change the formula's coefficients

### Changed

- Desktop notifications instead of beeping

### Removed

- Xfce plugin support


src/main.rs
#![windows_subsystem = "windows"]

use clap::{Parser, Subcommand};
use interprocess::local_socket::LocalSocketStream;

mod activities;
mod app;
mod math;

use app::{socket_name, App, Ipc};

/// Work-break balancer can track your work time and suggest break time
#[derive(Parser)]
pub struct Cli {
    #[command(subcommand)]
    pub command: Option<Commands>,
}

#[derive(Subcommand)]
pub enum Commands {
    /// Asks the app to reload configuration
    Reload,
    /// Sends you notification with the current status
    Notify,
    /// Switches between work and break time
    Switch,
    /// Terminates the app
    Terminate,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let cli = Cli::parse();

    let socket = socket_name();

    match &cli.command {
        Some(Commands::Reload) => {
            let stream = LocalSocketStream::connect(&*socket).map_err(|_| "App is not running")?;
            ron::ser::to_writer(stream, &Ipc::Reload)?;
        }
        Some(Commands::Notify) => {
            let stream = LocalSocketStream::connect(&*socket).map_err(|_| "App is not running")?;
            ron::ser::to_writer(stream, &Ipc::Notify)?;
        }
        Some(Commands::Switch) => {
            let stream = LocalSocketStream::connect(&*socket).map_err(|_| "App is not running")?;
            ron::ser::to_writer(stream, &Ipc::Switch)?;
        }
        Some(Commands::Terminate) => {
            let stream = LocalSocketStream::connect(&*socket).map_err(|_| "App is not running")?;
            ron::ser::to_writer(stream, &Ipc::Terminate)?;
        }
        None => {
            let mut app = App::new()?;
            if let Ok(stream) = LocalSocketStream::connect(&*socket) {
                ron::ser::to_writer(stream, &Ipc::Update)?;
                let (_, _, status) = app.status()?;

                println!("{}", status);
            } else {
                app.start()?;
            }
        }
    };

    Ok(())
}


src/math.rs
use std::time::Duration;

use serde::{Deserialize, Serialize};

pub struct Formula {
    a: f64,
    b: f64,
    c: f64,
}

impl Formula {
    pub fn new(a: &CoefficientA, b: &CoefficientB, c: &CoefficientC) -> Self {
        Self {
            a: a.0,
            b: b.0,
            c: c.0,
        }
    }
    pub fn compute_break(&self, strain: Duration) -> Duration {
        Duration::from_secs_f64(self.a * strain.as_secs_f64().powf(self.b) + self.c)
    }

    pub fn compute_strain(&self, work_break: Duration) -> Duration {
        Duration::from_secs_f64(
            (work_break
                .saturating_sub(Duration::from_secs_f64(self.c))
                .as_secs_f64()
                / self.a)
                .powf(self.b.recip()),
        )
    }
}

#[derive(Serialize, Deserialize)]
pub struct CoefficientA(pub f64);

impl Default for CoefficientA {
    fn default() -> Self {
        Self(0.00147884224225867)
    }
}

#[derive(Serialize, Deserialize)]
pub struct CoefficientB(pub f64);

impl Default for CoefficientB {
    fn default() -> Self {
        Self(1.67098454496329)
    }
}

#[derive(Serialize, Deserialize)]
pub struct CoefficientC(pub f64);

impl Default for CoefficientC {
    fn default() -> Self {
        Self(0.)
    }
}
#[cfg(test)]
mod tests {
    use std::time::Duration;

    use super::Formula;

    #[test]
    fn pomodoro_and_efficiency_conversions() {
        let formula = Formula::new(
            &Default::default(),
            &Default::default(),
            &Default::default(),
        );
        assert_eq! {formula.compute_break(Duration::from_secs(25 * 60)).as_secs_f64().round(), 5. * 60.};
        assert_eq! {formula.compute_strain(Duration::from_secs(5 * 60)).as_secs_f64().round(), 25. * 60.};

        assert_eq! {formula.compute_break(Duration::from_secs(52 * 60)).as_secs_f64().round(), 17. * 60.};
        assert_eq! {formula.compute_strain(Duration::from_secs(17 * 60)).as_secs_f64().round(), 52. * 60.};
    }
}


src/activities.rs
use std::time::{Duration, SystemTime};

use serde::{Deserialize, Serialize};

use crate::math::Formula;

#[derive(Serialize, Deserialize)]
pub struct Activity {
    pub start: SystemTime,
    pub end: Option<SystemTime>,
}

#[derive(Default, Serialize, Deserialize)]
pub struct Activities {
    pub list: Vec<Activity>,
}

impl Activities {
    pub fn truncate_until(&mut self, point: SystemTime) {
        self.list.retain_mut(|a| {
            a.end
                .map(|e| {
                    let retain = e > point;
                    if retain && a.start < point {
                        a.start = point
                    };
                    retain
                })
                .unwrap_or(true)
        });
    }

    /// Returns end, strain and work
    pub fn work(
        &self,
        formula: &Formula,
        now: SystemTime,
    ) -> (Option<SystemTime>, Duration, Duration) {
        let (end, strain, work) =
            self.list
                .iter()
                .fold((None, Duration::ZERO, Duration::ZERO), |state, item| {
                    let (prev_end, mut strain, mut total_work) = state;
                    let Activity { start, end } = item;
                    if let Some(prev_end) = prev_end {
                        let rest = start.duration_since(prev_end).unwrap();
                        strain = formula
                            .compute_strain(formula.compute_break(strain).saturating_sub(rest));
                    };
                    let work = end.unwrap_or(now).duration_since(*start).unwrap();
                    strain += work;
                    total_work += work;

                    (*end, strain, total_work)
                });
        (
            end,
            if let Some(end) = end {
                formula.compute_strain(
                    formula
                        .compute_break(strain)
                        .saturating_sub(now.duration_since(end).unwrap()),
                )
            } else {
                strain
            },
            work,
        )
    }

    pub fn switch(&mut self, now: SystemTime) {
        if let Some(last) = self.list.last_mut() {
            if last.end.is_some() {
                self.list.push(Activity {
                    start: now,
                    end: None,
                })
            } else {
                last.end = Some(now)
            }
        } else {
            self.list.push(Activity {
                start: now,
                end: None,
            })
        }
    }
}


src/app.rs
use std::{
    sync::mpsc::sync_channel,
    thread,
    time::{Duration, SystemTime, UNIX_EPOCH},
};

use chrono::{DateTime, Local};
use interprocess::local_socket::LocalSocketListener;
use notify_rust::{Notification, Urgency};
use serde::{Deserialize, Serialize};

use crate::{
    activities::Activities,
    math::{CoefficientA, CoefficientB, CoefficientC, Formula},
};

pub const APP_NAME: &str = env!("CARGO_PKG_NAME");

#[cfg(not(debug_assertions))]
pub const APP_UUID: &str = "19ef189a-0474-44a8-8168-530cc14ccacc";

#[cfg(debug_assertions)]
pub const APP_UUID: &str = "735d31a3-bfb0-47f9-9ce6-647202507738";

pub fn socket_name() -> String {
    format!("@{}_{}", APP_NAME, APP_UUID)
}

#[cfg(debug_assertions)]
pub const CONFIG_NAME: Option<&str> = Some("dev-config");

#[cfg(not(debug_assertions))]
pub const CONFIG_NAME: Option<&str> = None;

#[cfg(debug_assertions)]
pub const STATE_NAME: Option<&str> = Some("dev-state");

#[cfg(not(debug_assertions))]
pub const STATE_NAME: Option<&str> = Some("default-state");

#[derive(Default, Serialize, Deserialize)]
pub struct Config {
    #[serde(default)]
    pub coefficient_a: CoefficientA,
    #[serde(default)]
    pub coefficient_b: CoefficientB,
    #[serde(default)]
    pub coefficient_c: CoefficientC,
    #[serde(default)]
    pub daily_work_time_limit: u16,
    #[serde(default)]
    pub work_days_start_at: u8,
}
impl Config {
    pub fn load() -> Result<Self, Box<dyn std::error::Error>> {
        Ok(confy::load(APP_NAME, CONFIG_NAME)?)
    }
}

#[derive(Default, Serialize, Deserialize)]
pub struct State {
    #[serde(default)]
    pub activities: Activities,
}

#[derive(Serialize, Deserialize)]
pub enum Ipc {
    Reload,
    Update,
    Notify,
    Switch,
    Terminate,
}

pub struct App {
    last_strain: Duration,
    last_work: Duration,
    pub config: Config,
    state: State,
}

impl App {
    pub fn new() -> Result<App, Box<dyn std::error::Error>> {
        let config: Config = confy::load(APP_NAME, CONFIG_NAME)?;
        let state: State = confy::load(APP_NAME, STATE_NAME)?;
        let formula = Formula::new(
            &config.coefficient_a,
            &config.coefficient_b,
            &config.coefficient_c,
        );
        let now = SystemTime::now();
        let (_, last_strain, last_work) = state.activities.work(&formula, now);
        Ok(App {
            last_strain,
            last_work,
            config,
            state,
        })
    }

    pub fn status(&mut self) -> Result<(Duration, Duration, String), Box<dyn std::error::Error>> {
        let formula = Formula::new(
            &self.config.coefficient_a,
            &self.config.coefficient_b,
            &self.config.coefficient_c,
        );
        let now = SystemTime::now();
        let now_ch: DateTime<Local> = DateTime::from(now);
        let morning = now_ch
            .date_naive()
            .and_hms_opt(self.config.work_days_start_at.into(), 0, 0)
            .unwrap();
        let mut morning_local = morning.and_local_timezone(Local).unwrap();
        if now_ch < morning_local {
            morning_local = morning_local
                .checked_sub_days(chrono::Days::new(1))
                .unwrap();
        }
        let truncate_point = UNIX_EPOCH + Duration::from_secs(morning_local.timestamp() as u64);
        self.state.activities.truncate_until(truncate_point);

        let (end, strain, work) = self.state.activities.work(&formula, now);

        let phase = if !self.state.activities.list.is_empty() && end.is_none() {
            match strain.as_secs() / 60 {
                0..=24 => "Pomodoro",
                25..=51 => "Efficiency",
                52.. => "Injury",
            }
        } else if strain.is_zero() {
            "Ready"
        } else {
            "Break"
        };

        let status = format!(
            "Phase: {phase}\nStrain: {} min.\nBreak: {} min, ends at: {}\nToday: {} hrs, {} min.",
            strain.as_secs() / 60,
            formula.compute_break(strain).as_secs() / 60,
            DateTime::<Local>::from(now + formula.compute_break(strain)).format("%X"),
            work.as_secs() / 3600,
            (work.as_secs() - work.as_secs() / 3600 * 3600) / 60
        );

        Ok((strain, work, status))
    }

    pub fn notify(&mut self, notify_anyway: bool) -> Result<(), Box<dyn std::error::Error>> {
        let (strain, work, status) = self.status()?;

        const POMODORO: Duration = Duration::from_secs(25 * 60);
        const EFFICIENCY: Duration = Duration::from_secs(52 * 60);
        let daily_work_time_limit =
            Duration::from_secs(self.config.daily_work_time_limit as u64 * 60);
        let mut urgency = Urgency::Low;
        let notify_on_threshold = self.last_strain < POMODORO && strain >= POMODORO
            || self.last_strain < EFFICIENCY && strain >= EFFICIENCY
            || !self.last_strain.is_zero() && strain.is_zero()
            || !daily_work_time_limit.is_zero()
                && self.last_work < daily_work_time_limit
                && work >= daily_work_time_limit;

        if strain >= POMODORO {
            urgency = Urgency::Normal;
        }

        if strain >= EFFICIENCY || !daily_work_time_limit.is_zero() && work >= daily_work_time_limit
        {
            urgency = Urgency::Critical;
        }

        self.last_strain = strain;
        self.last_work = work;

        if notify_anyway || notify_on_threshold {
            let mut notification = Notification::new();
            notification
                .summary("Work-break balancer")
                .auto_icon()
                .body(&status);

            #[cfg(not(any(target_os = "windows", target_os = "macos")))]
            notification.urgency(urgency);

            #[cfg(target_os = "windows")]
            if urgency == Urgency::Critical {
                use notify_rust::Timeout;
                notification.timeout(Timeout::Never);
            };

            notification.show()?;
        }

        Ok(())
    }

    pub fn switch(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let now = SystemTime::now();
        self.state.activities.switch(now);
        confy::store(APP_NAME, STATE_NAME, &self.state)?;
        self.notify(true)?;
        Ok(())
    }

    pub fn start(&self) -> Result<(), Box<dyn std::error::Error>> {
        let socket = socket_name();

        let (sender, receiver) = sync_channel(0);

        let ctrlc_sender = sender.clone();
        ctrlc::set_handler(move || {
            ctrlc_sender.send(Ipc::Terminate).unwrap();
        })?;

        let socket_sender = sender.clone();
        thread::spawn(move || {
            let stream = LocalSocketListener::bind(socket).unwrap();
            println!("App has been started");

            for stream in stream.incoming() {
                if let Ok(stream) = stream {
                    if let Ok(ipc) = ron::de::from_reader(stream) {
                        socket_sender.send(ipc).unwrap();
                    } else {
                        eprintln!("Ipc message is not recognized!")
                    }
                } else {
                    socket_sender.send(Ipc::Terminate).unwrap();
                    break;
                }
            }
        });

        thread::spawn(move || loop {
            thread::park_timeout(Duration::from_secs(60));
            sender.send(Ipc::Update).unwrap();
        });

        let mut app = App::new()?;

        for ipc in receiver.iter() {
            match ipc {
                Ipc::Reload => {
                    app.config = Config::load()?;
                }
                Ipc::Update => {
                    app.notify(false)?;
                }
                Ipc::Notify => {
                    app.notify(true)?;
                }
                Ipc::Switch => {
                    app.switch()?;
                }
                Ipc::Terminate => {
                    break;
                }
            }
        }

        Ok(())
    }
}
